File: Linked List/LeetCode_24 Swap Nodes in Pairs/Swap Nodes in Pairs.c
struct ListNode* swapPairs(struct ListNode* head){
    
    struct ListNode *pGuardNode = (struct ListNode*)malloc(sizeof(struct ListNode)) ;
    struct ListNode*pre = pGuardNode;
    pGuardNode->next = head;
    while (pre->next && pre->next->next) {
        struct ListNode *t = pre->next->next;
        pre->next->next = t->next;
        t->next = pre->next;
        pre->next = t; /*in the first loop change pGuardNode->next*/
            
        pre = t->next;
    }
    return pGuardNode->next;

struct ListNode* swapPairs(struct ListNode* head){
    
    if (head == NULL || head->next == NULL) {
        return head;
    }
    struct ListNode *p = head->next;
    head->next = swapPairs(head->next->next);
    p->next = head;
    return p;
}

File: Linked List/LeetCode_24 Swap Nodes in Pairs/Swap Nodes in Pairs.py
class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        pre, pre.next = self, head
        while pre.next and pre.next.next:
            a = pre.next  #1st node
            b = a.next  #2nd node
            
            pre.next, b.next, a.next = b, a, b.next  #swap
            pre = a
            
        return self.next


class Solution:
    def swapPairs(self, head):
        if not head or not head.next:
            return head
        p = head.next
        head.next = self.swapPairs(head.next.next)
        p.next = head
        return p
